File: src/main/java/org/pl/interpreter/visitor/IfVisitor.java
package org.pl.interpreter.visitor;

import org.pl.interpreter.IInterpreter;
import org.pl.interpreter.exception.VisitorException;
import org.pl.parser.ast.INode;
import org.pl.parser.ast.IfCase;
import org.pl.parser.ast.IfNode;


/**
 * Interprets the conditional logic.
 */
public class IfVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return IfNode.class == node.getClass();
    }

    @Override
    public Object visit(IInterpreter interpreter, INode node) throws VisitorException {
        var ifNode = (IfNode) node;
        for (IfCase ifCase : ifNode.cases) {
            var conditionResult = interpreter.evalNode(ifCase.getCondition());
            if (Boolean.TRUE.equals(conditionResult)) {
                return interpreter.evalNode(ifCase.getBody());
            }
        }
        if (ifNode.elseCase != null) {
            return interpreter.evalNode(ifNode.elseCase);
        }
        return null;
    }
}

File: src/main/kotlin/org/leaflang/interpreter/visitor/IfVisitor.kt
package org.leaflang.interpreter.visitor

import org.leaflang.interpreter.IInterpreter
import org.leaflang.interpreter.result.IRuntimeResult
import org.leaflang.interpreter.result.emptyResult
import org.leaflang.parser.ast.IfNode
import org.leaflang.parser.ast.INode

/**
 * Interprets the conditional logic.
 */
class IfVisitor : IVisitor {
    override fun visit(interpreter: IInterpreter, node: INode): IRuntimeResult {
        val ifNode = node as IfNode
        for (ifCase in ifNode.cases) {
            val conditionResult = interpreter.interpret(ifCase.condition).data
            if (conditionResult?.value == true) {
                return interpreter.interpret(ifCase.body)
            }
        }
        if (ifNode.elseCase != null) return interpreter.interpret(ifNode.elseCase)
        return emptyResult()
    }
}